### PR TITLE
sql: implement date_bin builtin functions

### PR DESCRIPTION
**Pull request by sploiselle**
PG supports `date_bin`, and it would be nice to have for some tinkering I'm doing on timestamp windows #8698 (not to be totally conflated with window functions, which could be thought of as distinct).

### Motivation

This PR adds a known-desirable feature. #8866

### Tips for reviewer

The only real design decision here is adding a binary version of the function that defaults to using the Unix epoch as the `origin`. This could be a nice function for generating bins of specific widths whose origins simply need to be consistent, but don't need to be a particular value. Maybe we should wait to add this until PG does something, so as to not shoot ourselves in the foot with compatibility?

I also added `EvalError::FeatureNotSupported`, which feels useful for these kind of arbitrary input checks that you wouldn't want to create specific error codes for––but maybe this is too generic. Was inspired by PG's errors in the same implementation so ¯\\\_(ツ)\_/¯ @benesch You have a feeling about this?

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/expr/src/scalar/mod.rs [hunk 1]
```diff
@@ -1056,6 +1057,9 @@ impl fmt::Display for EvalError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
             EvalError::DivisionByZero => f.write_str("division by zero"),
+            EvalError::FeatureNotSupported(feature) => {
+                write!(f, "feature not supported: {}", feature)
```
Comment (benesch) [inline src/expr/src/scalar/mod.rs:1061]: Including "feature not supported" in the error text here seems maybe a bit weird. Doesn't seem to add any clarity to the error messages you've got so far, at least!

**Review by benesch — APPROVED** [review]
> Maybe we should wait to add this until PG does something, so as to not shoot ourselves in the foot with compatibility?

Yeah, probably. Maybe just mark the 2 arg version as experimental so that if we ever feel good about the tradeoff, we can stabilize it with basically zero effort?

**Comment by andrioni** [comment]
>Maybe we should wait to add this until PG does something, so as to not shoot ourselves in the foot with compatibility?

Do you know if they have any plans on that direction? It seems that this function was just added for Pg14, so I imagine they would have added it already if they wanted to.

File: src/expr/src/scalar/mod.rs (hunk 1 above)
Comment (andrioni) [reply src/expr/src/scalar/mod.rs:1061]: I would love if we had something Cockroach does where the error message actually links to a GitHub issue tracking support for that family of functions, [like here](https://github.com/cockroachdb/cockroach/blob/2e326b85fd0d60a62ae9d344b0be5c89e36bcc30/pkg/sql/sem/builtins/builtins.go#L3626).

File: doc/user/content/sql/functions/date-bin.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+---
+title: "date_bin Function"
+description: "Bins a timestamp into a specified interval"
+menu:
+  main:
+    parent: 'sql-functions'
+---
+
+`date_bin` returns the largest value less than `source` that is a multiple of
```
Comment (andrioni) [inline doc/user/content/sql/functions/date-bin.md:9]: less than or equal to?

File: doc/user/content/sql/functions/date-trunc.md [hunk 3]
```diff
@@ -6,19 +6,21 @@ menu:
     parent: 'sql-functions'
 ---
 
-`date_trunc` returns a [`timestamp`](../../types/timestamp) with the "floor value" of the specified time component, i.e. the largest time component less than or equal to the provided value.
+`date_trunc` returns a [`timestamp`] with the "floor value" of the specified time component, i.e. the largest time component less than or equal to the provided value.
```
Comment (andrioni) [inline doc/user/content/sql/functions/date-trunc.md:9]: Nitpick: Shouldn't this `timestamp` be replaced with `the same type as ts_val` as well? (or timestamp-like, or something equivalent)

File: doc/user/content/sql/functions/date-bin.md (hunk 2 above)
Comment (sploiselle) [reply doc/user/content/sql/functions/date-bin.md]: Good catch, ty

File: doc/user/content/sql/functions/date-trunc.md (hunk 3 above)
Comment (sploiselle) [reply doc/user/content/sql/functions/date-trunc.md]: Reworded

**Comment by sploiselle** [comment]
@benesch Put the binary implementation behind the experimental flag--great idea.

@andrioni I have an opposing expectation that they might not have known what an appropriate default origin was, so chose to not offer one, but they might come up with one in the next version after folks have had a chance to use it. That being said, I have no insight at all, so just erring on the side of caution.

**Pull request merged**